compiler/noirc_evaluator/src/ssa/opt/brillig_entry_points.rs

```diff
@@ -0,0 +1,612 @@
+        functions_to_clone_map.entry(entry_point).or_default().push((cloned_function, inner_call));
+    };
+
+    for (inner_call, inner_call_entry_points) in inner_call_to_entry_point {
+        if inner_call_entry_points.len() > 1 {
+            for entry_point in inner_call_entry_points {
+                add_function_to_clone(entry_point, inner_call);
+            }
+        } else if entry_points.contains(&inner_call) {
+            add_function_to_clone(inner_call_entry_points[0], inner_call);
+        }
+    }
+
+    functions_to_clone_map
+}
+
+// Clones new functions and returns a mapping representing the calls to update.
+// Returns a map of (entry point, callee function) -> new callee function id.
+fn build_calls_to_update(
+    ssa: &mut Ssa,
+    functions_to_clone_map: HashMap<FunctionId, Vec<(Function, FunctionId)>>,
+) -> HashMap<(FunctionId, FunctionId), FunctionId> {
+    let mut calls_to_update: HashMap<(FunctionId, FunctionId), FunctionId> = HashMap::default();
+
+    for (entry_point, functions_to_clone) in functions_to_clone_map {
+        for (mut cloned_function, old_id) in functions_to_clone {
+            ssa.add_fn(|id| {
+                calls_to_update.insert((entry_point, old_id), id);
+                cloned_function.set_id(id);
+                cloned_function
+            });
+        }
+    }
+
+    calls_to_update
+}
+
+fn update_function_calls(
+    function: &mut Function,
+    entry_point: FunctionId,
+    new_functions_map: &mut HashMap<FunctionId, FunctionId>,
+    // Maps (entry point, callee function) -> new callee function id
+    calls_to_update: &HashMap<(FunctionId, FunctionId), FunctionId>,
+) {
+    for block_id in function.reachable_blocks() {
+        #[allow(clippy::unnecessary_to_owned)] // clippy is wrong here
+        for instruction_id in function.dfg[block_id].instructions().to_vec() {
+            let instruction = function.dfg[instruction_id].clone();
+            let Instruction::Call { func: func_id, arguments } = instruction else {
+                continue;
+            };
+
+            let func_value = &function.dfg[func_id];
+            let Value::Function(func_id) = func_value else { continue };
+
+            if let Some(new_id) = calls_to_update.get(&(entry_point, *func_id)) {
+                new_functions_map.insert(*func_id, *new_id);
+                let new_function_value_id = function.dfg.import_function(*new_id);
+                function.dfg[instruction_id] =
+                    Instruction::Call { func: new_function_value_id, arguments: arguments.clone() };
+            } else {
+                continue;
+            }
+        }
+    }
+}
+
+/// Returns a map of Brillig entry points to all functions called in that entry point.
+/// This includes any nested calls as well, as we want to be able to associate
+/// any Brillig function with the appropriate global allocations.
+/// TODO: return a BTreeMap
+pub(crate) fn get_brillig_entry_points(
```

Comment:
This function must exist in like 3 other places at this point.

Reply:
I did deduplicate this logic from `brillig_globals.rs`, but that should be the only place?

Reply:
This was mostly based on a hunch as we've been building a lot of callgraph generating functions recently. No concrete example.

Feel free to resolve.